Stop updating modification time of origin file for every granted space request

szewai opened the pull request:
#### 93a78bc98d017d5b14d3fe642d9380347090c75b
<pre>
Stop updating modification time of origin file for every granted space request
<a href="https://bugs.webkit.org/show_bug.cgi?id=263416">https://bugs.webkit.org/show_bug.cgi?id=263416</a>
rdar://problem/117228957

Reviewed by Youenn Fablet.

We use last modification time of origin file to help decide data eviction ordering of origins, and we currently update
the time when a space request is granted for the origin. In real-world scenario, an origin might perform many storage
operations in a short period of time, leading to many last modification time updates. This is not ideal because the
timestamp may not change (or the change may not be big enough to affect eviction ordering), and updating modification
time is not free -- it could be slow when disk operation is involved, slowing down subsequent storage operations
on the same queue. To fix this, we now only update last modification time when at least 30 seconds have passed since
last update.

This change makes our existing IndexedDB tests run faster (`run-perf-tests IndexedDB/`).

* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::NetworkStorageManager::updateLastModificationTimeForOrigin):
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebsiteDatastore.mm:
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/269598@main">https://commits.webkit.org/269598@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5a32937abcd5fcdff9d10afd2abc6c812a11326d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/22998 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/1077 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/24112 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/24913 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/21278 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/23262 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/2309 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/23541 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/22148 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/23240 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/533 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/19950 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/25767 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/416 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/20833 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/27015 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/20797 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/21095 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/24875 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/515 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/18316 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/438 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5495 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/923 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/704 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -591,6 +592,17 @@ bool NetworkStorageManager::removeOriginStorageManagerIfPossible(const WebCore::
     assertIsCurrent(workQueue());
+
+    auto currentTime = WallTime::now();
+    auto lastModificationTime = m_lastModificationTimes.get(origin);
```
We could find and use an iterator so that we can optimise the set below.

youennf commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -591,6 +592,17 @@ bool NetworkStorageManager::removeOriginStorageManagerIfPossible(const WebCore::
+    });
+
     // This function must be called when origin is in use, i.e. OriginStorageManager exists.
     auto* manager = m_originStorageManagers.get(origin);
```
Why not storing lastModificationTime in manager instead?

szewai replied:
In current implementation we are actively removing OriginStorageManager when possible (see places `NetworkStorageManager::removeOriginStorageManagerIfPossible` is called), so there is a higher chance the `OriginStorageManager` goes away in the originLastModificationTimeUpdateInterval than `NetworkStorageManager`.

szewai committed:
Stop updating modification time of origin file for every granted space request
https://bugs.webkit.org/show_bug.cgi?id=263416
rdar://problem/117228957

Reviewed by Youenn Fablet.

We use last modification time of origin file to help decide data eviction ordering of origins, and we currently update
the time when a space request is granted for the origin. In real-world scenario, an origin might perform many storage
operations in a short period of time, leading to many last modification time updates. This is not ideal because the
timestamp may not change (or the change may not be big enough to affect eviction ordering), and updating modification
time is not free -- it could be slow when disk operation is involved, slowing down subsequent storage operations
on the same queue. To fix this, we now only update last modification time when at least 30 seconds have passed since
last update.

This change makes our existing IndexedDB tests run faster (`run-perf-tests IndexedDB/`).

* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::NetworkStorageManager::updateLastModificationTimeForOrigin):
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebsiteDatastore.mm:
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/269598@main
